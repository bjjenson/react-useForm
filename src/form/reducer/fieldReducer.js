import { List, Map } from 'immutable'
import { syncListIndexes } from './syncListIndexes'
import { mergeValidationResults } from './mergeValidationResults'
import { mergeFormValues } from '../helpers/mergeFormValues'

export const actionTypes = {
  insertField: 'insertField',
  removeField: 'removeField',
  addListItem: 'addListItem',
  removeListItem: 'removeListItem',
  updateListIndex: 'updateListIndex',
  reset: 'reset',
  touched: 'touched',
  updateValue: 'updateValue',
  validateAll: 'validateAll',
  validationResult: 'validationResult',
  addListener: 'addListener',
  removeListener: 'removeListener',
}

export const actions = {
  insertField: (fieldName, fieldState) => ({ type: actionTypes.insertField, fieldName, payload: fieldState }),
  removeField: fieldName => ({ type: actionTypes.removeField, fieldName }),
  addListItem: (fieldName, fieldState) => ({ type: actionTypes.addListItem, fieldName, payload: fieldState }),
  removeListItem: (fieldName, index) => ({ type: actionTypes.removeListItem, fieldName, payload: index }),
  updateListIndex: (fieldName, fromIndex, toIndex) => ({ type: actionTypes.updateListIndex, fieldName, payload: { fromIndex, toIndex } }),
  reset: (state) => ({ type: actionTypes.reset, payload: state }),
  touched: (fieldName) => ({ type: actionTypes.touched, fieldName }),
  updateValue: (fieldName, value) => ({ type: actionTypes.updateValue, fieldName, payload: value }),
  validateAll: (errors) => ({ type: actionTypes.validateAll, payload: errors }),
  validationResult: (fieldName, error, helperText) => ({ type: actionTypes.validationResult, fieldName, payload: { error, helperText } }),
  addListener: (fieldName, listener) => ({ type: actionTypes.addListener, fieldName, payload: listener }),
  removeListener: (fieldName, listener) => ({ type: actionTypes.removeListener, fieldName, payload: listener }),
}

export const removedFieldsKey = 'removedFields'
export const fieldsKey = 'fields'
export const listenersKey = 'listeners'

export const initState = state => {
  return state
}

const current = 'current'

export const fieldReducer = (state, { type, fieldName = '', payload }) => {
  const fieldPath = getFieldPath(fieldName)

  const handlers = {
    [actionTypes.updateValue]: value => {
      setTimeout(() => {
        const previous = state.getIn([fieldsKey, ...fieldPath, current, 'value'], '')
        state.getIn([listenersKey, fieldName], List()).forEach(listener => listener(value, previous, state.get('formTools').current))
      })

      const newState = state.setIn([fieldsKey, ...fieldPath, current, 'value'], value)
        .setIn([fieldsKey, ...fieldPath, current, 'pristine'], value == state.getIn([fieldsKey, fieldName, 'initial', 'value']))
      setTimeout(() => {
        const onFormChange = state.get('onFormChange')
        if (onFormChange && typeof (onFormChange) === 'function') {
          onFormChange(mergeFormValues(newState, Map()))
        }
      })

      return newState
    },

    [actionTypes.touched]: () =>
      state.setIn([fieldsKey, ...fieldPath, current, 'touched'], true),

    [actionTypes.validationResult]: ({ error, helperText }) => {
      return state.setIn([fieldsKey, ...fieldPath, current, 'error'], error)
        .setIn([fieldsKey, ...fieldPath, current, 'helperText'],
          helperText || state.getIn([fieldsKey, ...fieldPath, 'initial', 'field']).helperText
        )
    },

    [actionTypes.validateAll]: (errors) =>
      mergeValidationResults(state, errors, true),

    [actionTypes.insertField]: fieldState =>
      state.setIn([fieldsKey, ...fieldPath], fieldState)
        .deleteIn([removedFieldsKey, state.get(removedFieldsKey, List()).indexOf(fieldName)]),

    [actionTypes.removeField]: () =>
      state.deleteIn([fieldsKey, ...fieldPath])
        .set(removedFieldsKey, state.get(removedFieldsKey, List()).push(fieldName)),

    [actionTypes.addListItem]: itemState => {
      const path = [fieldsKey, ...fieldPath, 'items']
      return state.updateIn([...path], List(), prevState => prevState.push(itemState))
    },

    [actionTypes.removeListItem]: index =>
      syncListIndexes(
        state.deleteIn([fieldsKey, ...fieldPath, 'items', index]),
        fieldsKey,
        fieldPath,
      ),

    [actionTypes.updateListIndex]: ({ fromIndex, toIndex }) => {
      const path = [fieldsKey, ...fieldPath, 'items']
      const item = state.getIn([...path, fromIndex])
      return syncListIndexes(state.deleteIn([...path, fromIndex]).updateIn(path, items => items.insert(toIndex, item)), fieldsKey, fieldPath)
    },

    [actionTypes.addListener]: listener =>
      state.updateIn([listenersKey, fieldName], List(), listeners =>
        listeners.push(listener)
      ),

    [actionTypes.removeListener]: listener =>
      state.updateIn([listenersKey, fieldName], List(), listeners => {
        const index = listeners.indexOf(listener)
        return index > -1 ? listeners.delete(index) : listeners
      }),

    [actionTypes.reset]: state => state,
  }

  const handler = handlers[type]
  if (handler) {

    let lastPath = fieldPath
    if (!areArraysEqual(state.get('lastPath', []), lastPath)) {
      lastPath = []
    }

    const trackLastPath = type !== actionTypes.validationResult && type !== actionTypes.reset
    return handler(payload)
      .set('lastPath', trackLastPath ? lastPath : [])
  }
  return state
}

export const getFieldPath = (fieldName, prev = []) => {
  let fieldPath = [fieldName]
  const match = fieldName.match(/.items.(\d)./)

  if (match) {
    const firstField = fieldName.substr(0, match.index)
    const rest = fieldName.match(/.items.\d.fields.(.*)$/)[1]

    if (rest.match(/.items.(\d)./)) {
      return getFieldPath(rest, [firstField, 'items', match[1], 'fields'])
    }

    fieldPath = [...prev, firstField, 'items', match[1], 'fields', rest]

  }

  return fieldPath
}

const areArraysEqual = (a, b) => {
  if (a.length !== b.length) return false

  let same = true
  for (let i = 0; i < a.length; i++) {
    if (a[i] !== b[i]) { same = false }
  }
  return same
}
