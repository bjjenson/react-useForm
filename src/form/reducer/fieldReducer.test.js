import { fromJS } from 'immutable'
import { fieldReducer, actions, removedFieldsKey, getFieldPath } from './fieldReducer'
import { syncListIndexes } from './syncListIndexes'

jest.mock('./syncListIndexes')

jest.useFakeTimers()

let state, fieldName, formTools
beforeEach(() => {
  fieldName = 'field-name'
  formTools = {
    current: 'current-form-tools',
  }
  state = fromJS({
    fields: {
      [fieldName]: {
        initial: {
          value: '',
        },
        current: {
          value: '',
          pristine: true,
          touched: false,
          error: false,
          helperText: '',
        },
      },
      otherField: {
        initial: {
          value: '',
        },
        current: {
          value: '',
          pristine: true,
          touched: false,
          error: false,
          helperText: '',
        },
      },
    },
  }).set('formTools', formTools)

  syncListIndexes.mockImplementation(s => s)
  setTimeout.mockImplementation(cb => {
    cb()
  })
})

test('type not handled, return state', () => {
  expect(fieldReducer(state, {})).toEqual(state)
})

test('updateValue for field updates value on current', () => {
  const action = actions.updateValue(fieldName, 'new value')
  expect(fieldReducer(state, action)).toMatchSnapshot()
})

test('pristine is false if value not equal to initial', () => {
  const action = actions.updateValue(fieldName, 'new value')
  expect(fieldReducer(state, action)).toMatchSnapshot()
})

test('pristine is true if value equal to initial', () => {
  state = state.setIn([fieldName, 'current', 'pristine'], false)
  const action = actions.updateValue(fieldName, '')
  expect(fieldReducer(state, action)).toMatchSnapshot()
})

test('touched set on current field', () => {
  const action = actions.touched(fieldName)
  expect(fieldReducer(state, action)).toMatchSnapshot()
})

test('validationResult set on current field', () => {
  const action = actions.validationResult(fieldName, true, 'i am error')
  expect(fieldReducer(state, action)).toMatchSnapshot()
})

test('insertField', () => {
  const fieldState = fromJS({
    initial: {
      value: 'one',
    },
    current: {
      value: 'one',
    },
  })
  const action = actions.insertField('newField', fieldState)
  expect(fieldReducer(state, action)).toMatchSnapshot()
})

test('insertField removes fieldName from removedFields', () => {
  const fieldState = fromJS({
    initial: {
      value: 'one',
    },
    current: {
      value: 'one',
    },
  })
  const withRemoved = state.set(removedFieldsKey, fromJS(['newField']))
  const action = actions.insertField('newField', fieldState)
  expect(fieldReducer(withRemoved, action)).toMatchSnapshot()
})

test('removeField', () => {
  const action = actions.removeField(fieldName)
  expect(fieldReducer(state, action)).toMatchSnapshot()
})

describe('nested fields', () => {
  beforeEach(() => {
    state = fromJS({
      fields: {
        'parent.nested': {
          initial: {
            value: '',
          },
          current: {
            value: '',
            pristine: true,
            touched: false,
            error: false,
            helperText: '',
          },
        },
      },
    })
  })

  test('updateValue for field updates value on current', () => {
    const action = actions.updateValue('parent.nested', 'new value')
    expect(fieldReducer(state, action)).toMatchSnapshot()
  })

})

describe('list in state', () => {
  let listItemFieldName
  beforeEach(() => {
    listItemFieldName = `data.listField.items.1.fields.${fieldName}`
    state = fromJS({
      fields: {
        'data.listField': {
          items: [
            {},
            {
              fields: {
                [fieldName]: {
                  initial: {
                    value: '',
                  },
                  current: {
                    value: '',
                    pristine: true,
                    touched: false,
                    error: false,
                    helperText: '',
                  },
                },
              },
            },
          ],
        },
        otherField: {
          initial: {
            value: '',
          },
          current: {
            value: '',
            pristine: true,
            touched: false,
            error: false,
            helperText: '',
          },
        },
      },
    })
  })

  test('updateValue for field updates value on current', () => {
    const action = actions.updateValue(listItemFieldName, 'new value')
    expect(fieldReducer(state, action)).toMatchSnapshot()
  })

  test('touched set on current field', () => {
    const action = actions.touched(listItemFieldName)
    expect(fieldReducer(state, action)).toMatchSnapshot()
  })

  test('addListItem', () => {
    const fieldState = fromJS({
      fields: {
        [fieldName]: {
          initial: {
            type: 'text',
            value: '',
            optional: false,
            label: 'First',
            field: {
              label: 'First',
              name: 'data.listField.items.0.fields.fieldName',
            },
          },
          current: {
            helperText: '',
            error: false,
            pristine: true,
            touched: false,
            value: '',
          },
        },
      },
    })
    const action = actions.addListItem('data.listField', fieldState)
    expect(fieldReducer(state, action)).toMatchSnapshot()
  })

  test('removeListItem', () => {
    const action = actions.removeListItem('data.listField', 1)
    expect(fieldReducer(state, action)).toMatchSnapshot()
    expect(syncListIndexes.mock.calls[0]).toMatchSnapshot()
  })

  test('updateListIndex', () => {
    const action = actions.updateListIndex('data.listField', 0, 1)
    expect(fieldReducer(state, action)).toMatchSnapshot()
    expect(syncListIndexes.mock.calls[0]).toMatchSnapshot()
  })

})

describe('getFieldPath', () => {
  test('single name', () => {
    expect(getFieldPath('name')).toEqual(['name'])
  })

  test('combined name', () => {
    expect(getFieldPath('data.name')).toEqual(['data.name'])
  })

  test('single name with items', () => {
    expect(getFieldPath('listField.items.1.fields.fieldName')).toMatchSnapshot()
  })

  test('single name with items using combined name', () => {
    expect(getFieldPath('listField.items.1.fields.fieldName.sub1')).toMatchSnapshot()
  })

  test('combined name with items', () => {
    expect(getFieldPath('data.listField.items.1.fields.fieldName')).toMatchSnapshot()
  })

  test('list with combined names', () => {
    expect(getFieldPath('data.listField.items.0.fields.field1.sub1')).toMatchSnapshot()
  })

  test('list with more than 10 items', () => {
    expect(getFieldPath('data.listField.items.11.fields.field1.sub1')).toMatchSnapshot()
  })

  test('list of lists with single names', () => {
    expect(getFieldPath('data.listField.items.0.fields.field1.items.0.fields.field2')).toMatchSnapshot()
  })

  test('list of list with complex names', () => {
    expect(getFieldPath('data.listField.items.0.fields.field1.sub1.items.0.fields.field2.sub2')).toMatchSnapshot()
  })

  test('list of list with complex names longer than 10 items', () => {
    expect(getFieldPath('data.listField.items.11.fields.field1.sub1.items.11.fields.field2.sub2')).toMatchSnapshot()
  })
})

describe('fieldListener', () => {
  const listener = jest.fn()
  const listener2 = jest.fn()

  test('addListener sets listener into reducer', () => {
    const action = actions.addListener('name', listener)
    expect(fieldReducer(state, action)).toMatchSnapshot()
  })

  test('add multiple listeners', () => {
    const action = actions.addListener('name', listener)
    const action2 = actions.addListener('name', listener2)
    state = fieldReducer(state, action)
    state = fieldReducer(state, action2)

    expect(state).toMatchSnapshot()
  })

  test('listeners called when field value changes', () => {
    const previousValue = 'previousValue'
    state = state.setIn(['fields', fieldName, 'current', 'value'], previousValue)
    const update = 'i am update'
    const action = actions.addListener(fieldName, listener)
    const action2 = actions.addListener(fieldName, listener2)
    const changeAction = actions.updateValue(fieldName, update)

    state = fieldReducer(state, action)
    state = fieldReducer(state, action2)
    fieldReducer(state, changeAction)

    expect(listener).toHaveBeenCalledWith(update, previousValue, formTools.current)
    expect(listener2).toHaveBeenCalledWith(update, previousValue, formTools.current)
  })

  test('listeners called for nested field value changes', () => {
    const previousValue = 'previousValue'
    state = state.setIn(['fields', 'data', 'items', '0', 'fields', 'field', 'current', 'value'], previousValue)
    const update = 'i am update'
    const action = actions.addListener('data.field', listener)
    const changeAction = actions.updateValue('data.items.0.fields.field', update)

    state = fieldReducer(state, action)
    fieldReducer(state, changeAction)

    expect(listener).toHaveBeenCalledWith(update, previousValue, formTools.current)
  })

  test('listener not called for other fields', () => {
    const action = actions.addListener('name', listener)
    const changeAction = actions.updateValue('other', 'i am update')

    state = fieldReducer(state, action)
    state = fieldReducer(state, changeAction)

    expect(listener).not.toHaveBeenCalled()
  })

  test('removeListener removes by reference', () => {
    const previousValue = 'previousValue'
    state = state.setIn(['fields', fieldName, 'current', 'value'], previousValue)


    const update = 'i am update'
    const action = actions.addListener(fieldName, listener)
    const action2 = actions.addListener(fieldName, listener2)
    const removeAction = actions.removeListener(fieldName, listener)
    const changeAction = actions.updateValue(fieldName, update)

    state = fieldReducer(state, action)
    state = fieldReducer(state, action2)
    state = fieldReducer(state, removeAction)
    fieldReducer(state, changeAction)

    expect(state).toMatchSnapshot()
    expect(listener2).toHaveBeenCalledWith(update, previousValue, formTools.current)
    expect(listener).not.toHaveBeenCalled()
  })

})

test('validateAll merges validation errors', () => {
  const errors = {
    [fieldName]: 'error 1',
    otherField: 'error 2',
  }
  const action = actions.validateAll(errors)

  expect(fieldReducer(state, action)).toMatchSnapshot()
})

describe('onFormChange', () => {
  it('is called when present and a function', () => {
    const onFormChange = jest.fn()
    const action = actions.updateValue(fieldName, 'new value')
    const testState = state.set('onFormChange', onFormChange)
    fieldReducer(testState, action)
    expect(onFormChange).toHaveBeenCalledTimes(1)
    expect(onFormChange.mock.calls[0]).toMatchSnapshot()
  })
})

describe('updateFieldDefinition', ()=> {
  beforeEach(()=> {
    state = state.setIn(['fields', fieldName, 'initial', 'field'], {label: 'the-label'})
  })

  it('updateFieldDefinition updates optional', () => {
    const action = actions.updateFieldDefinition(fieldName, { definition: { name: fieldName, optional: true }, options: {} })
    expect(fieldReducer(state, action)).toMatchSnapshot()
  })

  it('updateFieldDefinition updates label', () => {
    const action = actions.updateFieldDefinition(fieldName, { definition: { name: fieldName, label: 'new-label' }, options: {} })
    expect(fieldReducer(state, action)).toMatchSnapshot()
  })

  it('updateFieldDefinition updates optional && label', () => {
    const action = actions.updateFieldDefinition(fieldName, { definition: { name: fieldName, optional: true, label: 'new-label' }, options: {} })
    expect(fieldReducer(state, action)).toMatchSnapshot()
  })

  it('updateFieldDefinition updates optional && label using formatter', () => {
    const action = actions.updateFieldDefinition(fieldName, { definition: { name: fieldName, optional: true, label: 'new-label' }, options: {optionalLabelFormatter: l=> `${l}***`} })
    expect(fieldReducer(state, action)).toMatchSnapshot()
  })

})
