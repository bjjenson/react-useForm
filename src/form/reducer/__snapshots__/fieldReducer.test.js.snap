// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fieldListener add multiple listeners 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "field-name": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
    "otherField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
  },
  "formTools": Object {
    "current": "current-form-tools",
  },
  "listeners": Immutable.Map {
    "name": Immutable.List [
      [MockFunction],
      [MockFunction],
    ],
  },
  "lastPath": Array [],
}
`;

exports[`fieldListener addListener sets listener into reducer 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "field-name": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
    "otherField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
  },
  "formTools": Object {
    "current": "current-form-tools",
  },
  "listeners": Immutable.Map {
    "name": Immutable.List [
      [MockFunction],
    ],
  },
  "lastPath": Array [],
}
`;

exports[`fieldListener removeListener removes by reference 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "field-name": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "previousValue",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
    "otherField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
  },
  "formTools": Object {
    "current": "current-form-tools",
  },
  "listeners": Immutable.Map {
    "field-name": Immutable.List [
      [MockFunction] {
        "calls": Array [
          Array [
            "i am update",
            "previousValue",
            "current-form-tools",
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      },
    ],
  },
  "lastPath": Array [],
}
`;

exports[`getFieldPath combined name with items 1`] = `
Array [
  "data.listField",
  "items",
  "1",
  "fields",
  "fieldName",
]
`;

exports[`getFieldPath list of list with complex names 1`] = `
Array [
  "data.listField",
  "items",
  "0",
  "fields",
  "field1.sub1",
  "items",
  "0",
  "fields",
  "field2.sub2",
]
`;

exports[`getFieldPath list of lists with single names 1`] = `
Array [
  "data.listField",
  "items",
  "0",
  "fields",
  "field1",
  "items",
  "0",
  "fields",
  "field2",
]
`;

exports[`getFieldPath list with combined names 1`] = `
Array [
  "data.listField",
  "items",
  "0",
  "fields",
  "field1.sub1",
]
`;

exports[`getFieldPath single name with items 1`] = `
Array [
  "listField",
  "items",
  "1",
  "fields",
  "fieldName",
]
`;

exports[`getFieldPath single name with items using combined name 1`] = `
Array [
  "listField",
  "items",
  "1",
  "fields",
  "fieldName.sub1",
]
`;

exports[`insertField 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "field-name": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
    "otherField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
    "newField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "one",
      },
      "current": Immutable.Map {
        "value": "one",
      },
    },
  },
  "formTools": Object {
    "current": "current-form-tools",
  },
  "lastPath": Array [],
}
`;

exports[`insertField removes fieldName from removedFields 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "field-name": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
    "otherField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
    "newField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "one",
      },
      "current": Immutable.Map {
        "value": "one",
      },
    },
  },
  "formTools": Object {
    "current": "current-form-tools",
  },
  "removedFields": Immutable.List [],
  "lastPath": Array [],
}
`;

exports[`list in state addListItem 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "data.listField": Immutable.Map {
      "items": Immutable.List [
        Immutable.Map {},
        Immutable.Map {
          "fields": Immutable.Map {
            "field-name": Immutable.Map {
              "initial": Immutable.Map {
                "value": "",
              },
              "current": Immutable.Map {
                "value": "",
                "pristine": true,
                "touched": false,
                "error": false,
                "helperText": "",
              },
            },
          },
        },
        Immutable.Map {
          "fields": Immutable.Map {
            "field-name": Immutable.Map {
              "initial": Immutable.Map {
                "type": "text",
                "value": "",
                "optional": false,
                "label": "First",
                "field": Immutable.Map {
                  "label": "First",
                  "name": "data.listField.items.0.fields.fieldName",
                },
              },
              "current": Immutable.Map {
                "helperText": "",
                "error": false,
                "pristine": true,
                "touched": false,
                "value": "",
              },
            },
          },
        },
      ],
    },
    "otherField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
  },
  "lastPath": Array [],
}
`;

exports[`list in state removeListItem 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "data.listField": Immutable.Map {
      "items": Immutable.List [
        Immutable.Map {},
      ],
    },
    "otherField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
  },
  "lastPath": Array [],
}
`;

exports[`list in state removeListItem 2`] = `
Array [
  Immutable.Map {
    "fields": Immutable.Map {
      "data.listField": Immutable.Map {
        "items": Immutable.List [
          Immutable.Map {},
        ],
      },
      "otherField": Immutable.Map {
        "initial": Immutable.Map {
          "value": "",
        },
        "current": Immutable.Map {
          "value": "",
          "pristine": true,
          "touched": false,
          "error": false,
          "helperText": "",
        },
      },
    },
  },
  "fields",
  Array [
    "data.listField",
  ],
]
`;

exports[`list in state touched set on current field 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "data.listField": Immutable.Map {
      "items": Immutable.List [
        Immutable.Map {},
        Immutable.Map {
          "fields": Immutable.Map {
            "field-name": Immutable.Map {
              "initial": Immutable.Map {
                "value": "",
              },
              "current": Immutable.Map {
                "value": "",
                "pristine": true,
                "touched": true,
                "error": false,
                "helperText": "",
              },
            },
          },
        },
      ],
    },
    "otherField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
  },
  "lastPath": Array [],
}
`;

exports[`list in state updateListIndex 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "data.listField": Immutable.Map {
      "items": Immutable.List [
        Immutable.Map {
          "fields": Immutable.Map {
            "field-name": Immutable.Map {
              "initial": Immutable.Map {
                "value": "",
              },
              "current": Immutable.Map {
                "value": "",
                "pristine": true,
                "touched": false,
                "error": false,
                "helperText": "",
              },
            },
          },
        },
        Immutable.Map {},
      ],
    },
    "otherField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
  },
  "lastPath": Array [],
}
`;

exports[`list in state updateListIndex 2`] = `
Array [
  Immutable.Map {
    "fields": Immutable.Map {
      "data.listField": Immutable.Map {
        "items": Immutable.List [
          Immutable.Map {
            "fields": Immutable.Map {
              "field-name": Immutable.Map {
                "initial": Immutable.Map {
                  "value": "",
                },
                "current": Immutable.Map {
                  "value": "",
                  "pristine": true,
                  "touched": false,
                  "error": false,
                  "helperText": "",
                },
              },
            },
          },
          Immutable.Map {},
        ],
      },
      "otherField": Immutable.Map {
        "initial": Immutable.Map {
          "value": "",
        },
        "current": Immutable.Map {
          "value": "",
          "pristine": true,
          "touched": false,
          "error": false,
          "helperText": "",
        },
      },
    },
  },
  "fields",
  Array [
    "data.listField",
  ],
]
`;

exports[`list in state updateValue for field updates value on current 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "data.listField": Immutable.Map {
      "items": Immutable.List [
        Immutable.Map {},
        Immutable.Map {
          "fields": Immutable.Map {
            "field-name": Immutable.Map {
              "initial": Immutable.Map {
                "value": "",
              },
              "current": Immutable.Map {
                "value": "new value",
                "pristine": false,
                "touched": false,
                "error": false,
                "helperText": "",
              },
            },
          },
        },
      ],
    },
    "otherField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
  },
  "lastPath": Array [],
}
`;

exports[`nested fields updateValue for field updates value on current 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "parent.nested": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "new value",
        "pristine": false,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
  },
  "lastPath": Array [],
}
`;

exports[`onFormChange is called when present and a function 1`] = `
Array [
  Immutable.Map {
    "field-name": "new value",
    "otherField": "",
  },
]
`;

exports[`pristine is false if value not equal to initial 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "field-name": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "new value",
        "pristine": false,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
    "otherField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
  },
  "formTools": Object {
    "current": "current-form-tools",
  },
  "lastPath": Array [],
}
`;

exports[`pristine is true if value equal to initial 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "field-name": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
    "otherField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
  },
  "formTools": Object {
    "current": "current-form-tools",
  },
  "field-name": Immutable.Map {
    "current": Immutable.Map {
      "pristine": false,
    },
  },
  "lastPath": Array [],
}
`;

exports[`removeField 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "otherField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
  },
  "formTools": Object {
    "current": "current-form-tools",
  },
  "removedFields": Immutable.List [
    "field-name",
  ],
  "lastPath": Array [],
}
`;

exports[`touched set on current field 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "field-name": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": true,
        "error": false,
        "helperText": "",
      },
    },
    "otherField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
  },
  "formTools": Object {
    "current": "current-form-tools",
  },
  "lastPath": Array [],
}
`;

exports[`updateValue for field updates value on current 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "field-name": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "new value",
        "pristine": false,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
    "otherField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
  },
  "formTools": Object {
    "current": "current-form-tools",
  },
  "lastPath": Array [],
}
`;

exports[`validateAll merges validation errors 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "field-name": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": true,
        "helperText": "error 1",
      },
    },
    "otherField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": true,
        "helperText": "error 2",
      },
    },
  },
  "formTools": Object {
    "current": "current-form-tools",
  },
  "lastPath": Array [],
}
`;

exports[`validationResult set on current field 1`] = `
Immutable.Map {
  "fields": Immutable.Map {
    "field-name": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": true,
        "helperText": "i am error",
      },
    },
    "otherField": Immutable.Map {
      "initial": Immutable.Map {
        "value": "",
      },
      "current": Immutable.Map {
        "value": "",
        "pristine": true,
        "touched": false,
        "error": false,
        "helperText": "",
      },
    },
  },
  "formTools": Object {
    "current": "current-form-tools",
  },
  "lastPath": Array [],
}
`;
